### PR TITLE
Document compatibility between modules

### PR DESCRIPTION
**Pull request by pellared**
Documents https://github.com/open-telemetry/opentelemetry-go/issues/3548

### CODE REVIEW

File: VERSIONING.md [hunk 1]
```diff
@@ -126,6 +126,12 @@ is designed so the following goals can be achieved.
       stable release except for a stable release of the contrib repository.
 * GitHub releases will be made for all releases.
 * Go modules will be made available at Go package mirrors.
+* This repository's modules are only guaranteed to be compatible with
+  modules released from the same commit. For example,
```
Comment (MrAlias) [inline VERSIONING.md:130]: It should be clear here that we are talking cross module compatibility by using the term "other modules" and we want to guarantee across the release not a commit.

```suggestion
* Go modules are only guaranteed to be compatible with
  other modules from the same release. They are not
  guaranteed to be compatible with other modules from
  previous releases.  For example,
```

File: VERSIONING.md [hunk 2]
```diff
@@ -126,6 +126,12 @@ is designed so the following goals can be achieved.
...
+  `go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.11.1` is not
+  compatible with `go.opentelemetry.io/otel v1.11.2`.
+  Tracked under:
+  [#3548](https://github.com/open-telemetry/opentelemetry-go/issues/3548).
```
Comment (MrAlias) [inline VERSIONING.md:134]: I don't think it is necessary to include an issue link here. The git blame will show this PR which links to the origin if a dev needs to know.

```suggestion
```

**Review by MrAlias — COMMENTED** [review]
Overall looks good. Thanks for adding this.

**Review by Aneurysm9 — CHANGES_REQUESTED** [review]
I'm not sure this is the correct approach here.  I understand it is what we have done to this point and expect to continue doing, but I think it is in conflict with [the specification's versioning guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#design-goals).

> Never create a dependency conflict between packages which rely on different versions of OpenTelemetry.

It is possible (maybe even likely) that an instrumentation library will require a newer version of the OTel API while the application still requires an older version of the SDK.  This puts the user in the position of needing to choose between updating an instrumented dependency or staying on a known-working version of the SDK.

What options do we have for avoiding this coupling?  I think mutual dependencies from the API back to the SDK modules that depend on its internal packages is a non-starter.  What about managed duplication through code generation?  Are there other approaches we could take?

**Comment by pellared** [comment]
> What about managed duplication through code generation?

That is exactly what I propose [here](https://github.com/open-telemetry/opentelemetry-go/issues/3548#issuecomment-1443172592) 

> We could address the problem e.g. by using "code generation" to "copy/paste" the common code that is shared between modules.

Having that said, I think we should document the current state before we address the issue.

**Comment by ericklaus-wf** [comment]
I think the situation is more complicated than described in this documentation update. It isn't the case that modules are only compatible at the same version (`go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.11.1` is only compatible with `go.opentelemetry.io/otel v1.11.1`). Instead, modules are only guaranteed compatible at releases _from the same commit_, which _may be different versions_.

For example, `go.opentelemetry.io/otel/exporters/otlp/otlpmetric@v0.32.1` is compatible with `go.opentelemetry.io/otel@v1.11.1` but is not compatible with `go.opentelemetry.io/otel@v1.11.2`. The only way to discover compatible versions is to manually inspect each git tag, looking for tags at the same commit. That feels like a pretty onerous burden, but documenting it here will at least make it clearer to your integrators what they're signing up for.

**Comment by MadVikingGod** [comment]
@pellared I did a bit of work to identify which modules take dependencies on cross-module `internal` packages [here](https://github.com/open-telemetry/opentelemetry-go/issues/3548#issuecomment-1452106454).

**Comment by pellared** [comment]
Closing. I assigned https://github.com/open-telemetry/opentelemetry-go/issues/3548 to myself.

**Comment by pellared** [comment]
> I think the situation is more complicated than described in this documentation update. It isn't the case that modules are only compatible at the same version (`go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.11.1` is only compatible with `go.opentelemetry.io/otel v1.11.1`). Instead, modules are only guaranteed compatible at releases _from the same commit_, which _may be different versions_.
> 
> For example, `go.opentelemetry.io/otel/exporters/otlp/otlpmetric@v0.32.1` is compatible with `go.opentelemetry.io/otel@v1.11.1` but is not compatible with `go.opentelemetry.io/otel@v1.11.2`. The only way to discover compatible versions is to manually inspect each git tag, looking for tags at the same commit. That feels like a pretty onerous burden, but documenting it here will at least make it clearer to your integrators what they're signing up for.

Correct. https://github.com/open-telemetry/opentelemetry-go/pull/3806/commits/cf20bc3bbfc289b93478215f580306223fe99781 😉

**Pull request closed without merging**